Add OpenTelemetry Collector extension/component

martinjt opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠️ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! -->

<!-- ⚠️ **Every PR** needs to have a linked issue and have previously been approved. PRs that don't follow this will be rejected. >

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Closes #602 **

<!-- Add an overview of the changes here -->

<!-- All details should be in the linked issue. Feel free to call out any outstanding differences here. -->

## PR Checklist

<!-- Please check if your PR fulfills the following requirements, and remove the ones that are not applicable to the current PR -->

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] New integration
  - [x] Docs are written
  - [x] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

## Other information

<!-- Please add any other information that might be helpful to reviewers. -->

martinjt commented:
@dotnet-policy-service agree

martinjt commented:
I'm not sure how ready this is yet. I've been playing with the API a bit to give it enough flexibility around the certificates, images and config files.

Feedback very welcome, I wanted to get this up as is before I added any tests and worked on the documentation, etc.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector.csproj:
```diff
@@ -0,0 +1,12 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net9.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
```
```suggestion
```

This will all be provided by the props files above.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableHook.cs:
```diff
@@ -0,0 +1,53 @@
+{
+    private readonly ILogger<EnvironmentVariableHook> _logger;
+
+    public EnvironmentVariableHook(ILogger<EnvironmentVariableHook> logger)
+    {
+        _logger = logger;
+    }
```
Use a primary constructor

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableHook.cs:
```diff
@@ -0,0 +1,53 @@
+        var resources = appModel.GetProjectResources();
+        var collectorResource = appModel.Resources.OfType<CollectorResource>().FirstOrDefault();
+
+        if (collectorResource == null)
```
```suggestion
        if (collectorResource is null)
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableHook.cs:
```diff
@@ -0,0 +1,53 @@
+            return Task.CompletedTask;
+        }
+
+        if (resources.Count() == 0)
```
```suggestion
        if (!resources.Any())
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableHook.cs:
```diff
@@ -0,0 +1,53 @@
+        foreach (var resourceItem in resources)
+        {
+            _logger.LogDebug($"Forwarding Telemetry for {resourceItem.Name} to the collector");
+            if (resourceItem == null) continue;
```
```suggestion
            if (resourceItem is null) continue;
```

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/PublicAPI.Shipped.txt:
File no longer needed

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/PublicAPI.Unshipped.txt:
File no longer needed

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/DevCertHostingExtensions.cs:
We can likely use this in a few places, it should help unblock #388 and #444, so could it go into the `Shared` folder and be added to the csproj.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/OpenTelemetryCollectorSettings.cs:
```diff
@@ -0,0 +1,27 @@
+    /// <summary>
+    /// The version of the collector, defaults to latest
+    /// </summary>
+    public string CollectorVersion { get; set; } = "latest";
```
We should pin to a version so it's known stable per the package version

martinjt replied:
I'm not convinced that pinning is the right approach for that, especially since there isn't a v1 yet. It's generally accepted that people don't pin even in production.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/OpenTelemetryCollectorSettings.cs:
```diff
@@ -0,0 +1,27 @@
+    /// <summary>
+    /// The image of the collector, defaults to ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-contrib
+    /// </summary>
+    public string CollectorImage { get; set; } = "ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-contrib";
```
I feel like this should be split to registry/image like we do with other container resources. And is there a reason you're using ghcr.io not Docker Hub?

martinjt replied:
ghcr is the recommended default since the limits went into dockerhub.

Do you have an example of the pattern other integrations use for the registry split.

aaronpowell replied:
https://github.com/CommunityToolkit/Aspire/blob/0c0afa121b5c4078310f1f13eca49cbe3c77c609/src/CommunityToolkit.Aspire.Hosting.Ollama/OllamaContainerImageTags.cs#L5-L6

aaronpowell reviewed (commented):
Would it make sense that, rather than having a default collector image set, that it's a required property to be provided and there's a method that will set the defaults for you?

So, `AddOpenTelemetryCollector` is used if you want to specify a collector, whereas `AddOpenTelemetryDefaultCollection` uses the one that you have currently specified as a default.

martinjt commented:
Interesting approach, I'm honestly not sure, but maybe a `.WithDefaultImage()` might be better?

I was thinking that the approach I showed was more common, but happy if not.

I kinda wanted a minimal default command though.

aaronpowell commented:
> Interesting approach, I'm honestly not sure, but maybe a `.WithDefaultImage()` might be better?
> 
> I was thinking that the approach I showed was more common, but happy if not.
> 
> I kinda wanted a minimal default command though.

The problem with a method like `WithDefaultImage` is that you could setup one without any image specified which would be a runtime error, which would be a less than desirable outcome.

martinjt commented:
I agree it's not ideal, but I'd prefer to not have them need to know what the image link is to collector. So I believe the best option is to use the default image we recommend in Otel right now (which is `latest`) when the extension is called without parameters.

martinjt commented:
I'm coming back at this now @aaronpowell, with a hope to move it forward so I don't need to maintain a separate one (happy to be codeowner here).

On the devcerts, it seems those issues haven't got a resolution, would it be ok to leave the devcerts in this project for now, and consolidate/move later? This project would actually fix both of those issues as the users could run the non-ssl collector and forward their data through the collector as a way to unblock them.

The big outstanding issue is the collector image model, can you let me know what you'd suggest there. It should be `latest` since there isn't a v1 of the collector yet, and the advice from the collector team is to always use latest.

aaronpowell commented:
> On the devcerts, it seems those issues haven't got a resolution, would it be ok to leave the devcerts in this project for now, and consolidate/move later? This project would actually fix both of those issues as the users could run the non-ssl collector and forward their data through the collector as a way to unblock them.

If you can pop it in the [`Shared` folder](https://github.com/CommunityToolkit/Aspire/tree/main/src/Shared) we can then add it as a linked file into several projects that could benefit from it in the CT as well.

> The big outstanding issue is the collector image model, can you let me know what you'd suggest there. It should be `latest` since there isn't a v1 of the collector yet, and the advice from the collector team is to always use latest.

What's the release frequency that they are going with? The main reason that we pin to a version is so that we don't risk any unexpected changes to command line args or other behaviour getting introduced that impact someone who hasn't opt-ed in to using a newer version. If they are releasing new versions daily, then pinned isn't ideal, but generally speaking, pinning a version is going to be preferred, and people can always specify a version override if we don't release an updated package quick enough (Ollama is a good example of where we get caught on that front).

martinjt commented:
The release cycle for the collector is every 2 weeks. The compatibility issues wouldn't be commandline args, its more likely the config of collector components that might change, but we haven't done a breaking change in about 6 months and those are expected to be the last.

The biggest issue with pinning a version is security issues, since the amount of components in the collector means that the updates contain those updates regularly.

I'll happily take guidance on what the best way to get this published is.

martinjt commented:
I've done a few major updates, so it probably needs a full re-review.

* Implemented inbound SSL, which can also be forced off (this helps fix some of the clients that can't use HTTPS because of the certificate issues)
* Outbound SSL was removed, it won't work until the Certificate supports `host.docker.internal` or the `*.localhost` stuff works.
* Moved to using `.WithConfig("config.yaml")`, I'm not sure how I'd fail startup if they don't provide that, I'm happy to implement it.
* `.WithConfig()` supports multiple configs
* Moved to using Random ports supplied by Aspire
* Added SSL certificates for HTTP and gRPC
* Made the Environment variables detect the protocol that the App wants and use that (useful as Node, by default, supports HTTP not gRPC)
* Support for both of the endpoint environment variable names `ASPIRE_DASHBOARD_OTLP_ENDPOINT_URL` and `DOTNET_DASHBOARD_OTLP_ENDPOINT_URL`
* Updated DevCerts to not try and add the certificates twice (this will only be useful when the Outbound SSL functionality is available.

This does require the user to implement `insecure_skip_verify`, but I can't find a way around that until the new certificate work to make the authority match.

davidfowl commented on src/Shared/DevCertHostingExtensions.cs:
```diff
@@ -0,0 +1,199 @@
+    /// This method <strong>does not</strong> configure an HTTPS endpoint on the resource.
+    /// Use <see cref="ResourceBuilderExtensions.WithHttpsEndpoint{TResource}"/> to configure an HTTPS endpoint.
+    /// </remarks>
+    public static IResourceBuilder<TResource> RunWithHttpsDevCertificate<TResource>(
```
We don't ship this in any of the intergrations as yet and I think we'd rewrite it at this point to behave a bit better given the advances in the model.

aaronpowell replied:
Is there something "wrong" in the design of how this has been implemented that should be revised? Having a way to bootstrap the dev certificate into a container has been a pain point with a few integrations and having them working with OTEL (DAB and Go Feature Flags off the top of my head).

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/CollectorExtensions.cs:
```diff
@@ -0,0 +1,105 @@
+    /// <returns></returns>
+    public static IResourceBuilder<CollectorResource> WithAppForwarding(this IResourceBuilder<CollectorResource> builder)
+    {
+        builder.ApplicationBuilder.Services.TryAddLifecycleHook<EnvironmentVariableHook>();
```
Lifecycle hooks are mostly considered legacy with the event system now in place. The hook should be replacable with the `OnAfterEndpointsAllocated` callback.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableHook.cs:
```diff
@@ -0,0 +1,59 @@
+            return Task.CompletedTask;
+        }
+
+        var grpcEndpoint = collectorResource!.GetEndpoint(collectorResource!.GRPCEndpoint.EndpointName);
```
Shouldn't need to force non-null here, the `if` clause above should catch that.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableHook.cs:
```diff
@@ -0,0 +1,59 @@
+                var protocol = context.EnvironmentVariables.GetValueOrDefault("OTEL_EXPORTER_OTLP_PROTOCOL", "");
+                var endpoint = protocol.ToString() == "http/protobuf" ? httpEndpoint : grpcEndpoint;
+
+                if (endpoint == null)
```
```suggestion
                if (endpoint is null)
```

aaronpowell commented on src/Shared/DevCertHostingExtensions.cs:
```diff
@@ -0,0 +1,199 @@
+    {
+        if (builder.ApplicationBuilder.ExecutionContext.IsRunMode && builder.ApplicationBuilder.Environment.IsDevelopment())
+        {
+            builder.ApplicationBuilder.Eventing.Subscribe<BeforeStartEvent>(async (e, ct) =>
```
Let's use the [`OnBeforeStart`](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.lifecycle.idistributedapplicationlifecyclehook.beforestartasync?view=dotnet-aspire-9.0) callback as it is a cleaner API

aaronpowell reviewed (changes requested):
The main change here would be to use the event callbacks of 9.4.

Then would you be able to introduce an example project so I can get a better idea of how this works.

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableHook.cs:
```diff
@@ -0,0 +1,59 @@
+/// Hooks to add the OTLP environment variables to the various containers
+/// </summary>
+/// <param name="logger"></param>
+public class EnvironmentVariableHook(ILogger<EnvironmentVariableHook> logger) : IDistributedApplicationLifecycleHook
```
On top of moving this implementation to the eventing model, we should also use this [OtlpExporterAnnotation](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.applicationmodel.otlpexporterannotation?view=dotnet-aspire-9.0) to detect which resources should be impacted.

aaronpowell commented:
Yes, sorry, I was mistaken, `OnResourceEndpointsAllocated` is the event that you'd want on the collector resource.

Is the logic that you'd have multiple collectors registered, or is it a singleton sort of thing?

martinjt commented:
Other than tests, I think this is pretty much there now @aaronpowell @davidfowl 

I've struggled with the tests because the main thing to test is the AppForwarding (testing that all OTHER resources have their OTEL endpoint environment variables updated to point to the collector). This means that the endpoints need to be allocated in the collector before the event is fired.

The only other tests I can think of that would be beneficial are the environment variables for the Dashboard endpoints, and the devcerts, I'm not sure how we can test those. 

If you've got pointers on the tests then please let me know!

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.OpenTelemetryCollector/EnvironmentVariableEvent.cs:
```diff
@@ -0,0 +1,95 @@
+/// <summary>
+/// Hooks to add the OTLP environment variables to the various resources
+/// </summary>
+public static class EnvironmentVariableEventExtention
```
These methods should be internal.

martinjt commented:
Looks like it should all be resolved now.

aaronpowell commented:
Tests need to be added to the workflow file, since we run a job-per-test. Check https://github.com/CommunityToolkit/Aspire/actions/runs/17362288128/job/49289169742?pr=603#step:4:11
